### PR TITLE
Make contribution types configurable

### PR DESCRIPTION
**Pull request by tomrf1**
## Why are you doing this?
Contribution types (aka frequencies) are now configurable from the Support Admin Console.
This lets us configure which contributionTypes are available in each region, and which is the default.

This change gets the contributionTypes settings from s3 (currently they're in the javascript) and sends them to the client.

## Changes

* Server gets the settings from `contributionTypes.json` (along with switches and amounts settings)
* Client uses the settings in window.guardian.settings

### CODE REVIEW

File: support-frontend/assets/helpers/globals.js [hunk 1]
```diff
@@ -16,46 +16,33 @@ function getGlobal<T>(path: string = ''): ?T {
...
+    GBPCountries: emptyAmountsSettings,
+    UnitedStates: emptyAmountsSettings,
+    EURCountries: emptyAmountsSettings,
+    AUDCountries: emptyAmountsSettings,
+    International: emptyAmountsSettings,
+    NZDCountries: emptyAmountsSettings,
+    Canada: emptyAmountsSettings,
+  },
+  contributionTypes: {
+    GBPCountries: [],
```
Comment (tomrf1) [inline support-frontend/assets/helpers/globals.js]: I wonder if it's worth combining the amounts + contributionTypes settings into one model, e.g.

```
type Amount = {
  value: string,
  isDefault?: boolean,
}

type ContributionTypeSetting = {
  contributionType: ContributionType,
  isDefault?: boolean,
  amounts: Amounts[],
}

type ContributionTypesSettings = {
  [CountryGroupId]: ContributionTypeSetting[],
}
```

File: support-frontend/app/admin/settings/ContributionTypes.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+sealed trait ContributionType
+case object ONE_OFF extends ContributionType
+case object MONTHLY extends ContributionType
+case object ANNUAL extends ContributionType
+
+case object ContributionType {
+  implicit val contributionTypeEncoder: Encoder[ContributionType] = Encoder.encodeString.contramap[ContributionType](_.toString)
+
+  implicit val contributionTypeDecoder: Decoder[ContributionType] = JsonHelpers.decodeStringAndCollect {
```
Comment (joelochlann) [inline support-frontend/app/admin/settings/ContributionTypes.scala]: is there a circe macro that could avoid this boilerplate? or maybe using enumeratum? just curious!

Comment (tomrf1) [reply support-frontend/app/admin/settings/ContributionTypes.scala]: good suggestion - looks like enumeratum solves this https://github.com/lloydmeta/enumeratum#circe
([explanation of why it doesn't do this out of the box](https://stackoverflow.com/a/37015184))

File: support-models/src/main/scala/com/gu/support/encoding/JsonHelpers.scala [hunk 3]
```diff
@@ -98,4 +98,11 @@ object JsonHelpers {
     def getField(key: String) = json.hcursor.downField(key).focus
   }
 
+  //Decodes an expected json string to T if the given partial function is defined for the string value
+  def decodeStringAndCollect[T](pf: PartialFunction[String,T]): Decoder[T] = Decoder.decodeString.emap { s =>
```
Comment (joelochlann) [inline support-models/src/main/scala/com/gu/support/encoding/JsonHelpers.scala]: is this what works in conjunction with the stuff in `ContributionTypes.scala` to encode/decode a sum type to JSON? (asking for my own Scala curiosity!)

File: support-frontend/assets/pages/new-contributions-landing/contributionsLandingInit.js [hunk 4]
```diff
@@ -63,13 +64,25 @@ function getInitialPaymentMethod(
...
+      contributionTypes[countryGroupId][0].contributionType;
+  };
+
+  const contributionType =
+    getContributionTypeFromUrlOrElse(getContributionTypeFromSessionOrElse(getDefaultContributionType()));
+
   return (
     // make sure we don't select a contribution type which isn't on the page
-    getValidContributionTypes(countryGroupId).includes(contributionType)
+    contributionTypes[countryGroupId].find(ct => ct.contributionType === contributionType)
```
Comment (joelochlann) [inline support-frontend/assets/pages/new-contributions-landing/contributionsLandingInit.js]: the overall logic of this function is a bit hard to follow for me now. especially the appearance of the `contributionTypes[countryGroupId][0].contributionType` fallback in two places - is there a way to clarify?

File: support-frontend/assets/helpers/globals.js (hunk 1 above)
Comment (joelochlann) [reply support-frontend/assets/helpers/globals.js]: Yes! We could put everything that's specific to region + contribution type in there.

Comment (joelochlann) [reply support-frontend/assets/helpers/globals.js]: (e.g. possibly the "editorialise amounts" copy, `Contributing £50 works out as 14p a day`)

File: support-models/src/main/scala/com/gu/support/encoding/JsonHelpers.scala (hunk 3 above)
Comment (tomrf1) [reply support-models/src/main/scala/com/gu/support/encoding/JsonHelpers.scala]: yep, it's just a convenient way to match on a json string, for case object decoding

File: support-frontend/assets/pages/new-contributions-landing/components/ContributionTypeTabs.jsx [hunk 5]
```diff
@@ -18,6 +17,7 @@ import type { Switches } from 'helpers/settings';
 import type { CountryGroupId } from 'helpers/internationalisation/countryGroup';
 import { type State } from '../contributionsLandingReducer';
 import { updateContributionTypeAndPaymentMethod } from '../contributionsLandingActions';
+import type { ContributionTypes, ContributionTypeSetting } from '../../../helpers/contributions';
```
Comment (joelochlann) [inline support-frontend/assets/pages/new-contributions-landing/components/ContributionTypeTabs.jsx]: I think I'd rather not have a relative path to `helpers` in case the file moves... I know that there are other relative paths, but they're to very local contribution-specific files rather than the global helpers

File: support-frontend/assets/pages/new-contributions-landing/contributionsLandingInit.js [hunk 6]
```diff
@@ -45,6 +45,7 @@ import type { ExistingPaymentMethod } from '../../helpers/existingPaymentMethods
 import { setExistingPaymentMethods } from '../../helpers/page/commonActions';
 import { doesUserAppearToBeSignedIn } from '../../helpers/user/user';
 import { isSwitchOn } from 'helpers/globals';
+import type { ContributionTypes } from '../../helpers/contributions';
```
Comment (joelochlann) [inline support-frontend/assets/pages/new-contributions-landing/contributionsLandingInit.js]: ahh - I see lots more already existing relative paths to `helpers` here!

File: support-frontend/test/codecs/CirceDecodersTest.scala [hunk 7]
```diff
@@ -192,7 +192,51 @@ class CirceDecodersTest extends WordSpec with MustMatchers {
           |            { "value": "25", "isDefault": true }
           |        ]
           |    }
-          |  }
+          |  },
+          |  "contributionTypes": {
+          |"GBPCountries": [
```
Comment (joelochlann) [inline support-frontend/test/codecs/CirceDecodersTest.scala]: I know it's a bit of a pain but it would be nice to have correct indentation here to help skim-reading the file

File: support-frontend/assets/helpers/globals.js (hunk 1 above)
Comment (tomrf1) [reply support-frontend/assets/helpers/globals.js]: the migration could be quite complicated so I'll create a trello card to address this

File: support-frontend/assets/pages/new-contributions-landing/contributionsLandingInit.js (hunk 4 above)
Comment (tomrf1) [reply support-frontend/assets/pages/new-contributions-landing/contributionsLandingInit.js]: I've just pushed https://github.com/guardian/support-frontend/pull/1758/commits/3c3495042def3dfdc794905cc47b12427e4d7f5b to attempt to address this.
It checks the url/session (which may or may not be valid) before using the settings.
